klippy/extras/filament_switch_sensor.py

```diff
@@ -0,0 +1,165 @@
+        self.runout_cb = self._runout_event_handler
+        self.insert_cb = self._autoload_event_handler
+        self.runout_enabled = False
+        self.insert_enabled = False
+        self.event_running = False
+        self.monitor_state = True
+        self.print_status = "idle"
+        for status in ["idle", "ready", "printing"]:
+            self.printer.register_event_handler(
+                "idle_timeout:%s" % (status),
+                (lambda e, s=self, st=status: s._update_print_status(st)))
+        self.printer.register_event_handler("klippy:ready", self._handle_ready)
+        self.gcode.register_mux_command(
+            "QUERY_FILAMENT_SENSOR", "SENSOR", self.name,
+            self.cmd_QUERY_FILAMENT_SENSOR,
+            desc=self.cmd_QUERY_FILAMENT_SENSOR_help)
+    def _handle_ready(self):
+        self.toolhead = self.printer.lookup_object('toolhead')
+        # delay turning on insert detection a bit.  This prevents an
+        # insert event triggering on startup
+        if self.autoload_gcode is not None:
+            reactor = self.printer.get_reactor()
+            reactor.register_callback(
+                (lambda e, s=self, st=self.print_status:
+                 s._update_print_status(st)),
+                reactor.monotonic() + 2.)
+    def _update_print_status(self, status):
+        logging.info(
+            "filament_sensor: print status changed to: %s" % (status))
+        self.print_status = status
+        if not self.monitor_state:
+            return
+        if status == "printing":
+            runout_en = self.runout_gcode is not None
+            self.set_enable(runout_en, False)
+        else:
+            insert_en = self.autoload_gcode is not None
+            self.set_enable(False, insert_en)
+    def _runout_event_handler(self, eventtime):
+        if self.event_running:
+            return
+        self.event_running = True
+        # Pausing from inside an event requires that the pause portion
+        # of pause_resume execute immediately.
+        pause_resume = self.printer.lookup_object('pause_resume', None)
+        if self.runout_pause and pause_resume is not None:
+            pause_resume.send_pause_command(from_event=True)
+        self._exec_gcode(self.runout_gcode)
+        self.event_running = False
+    def _autoload_event_handler(self, eventtime):
+        if self.event_running:
+            return
+        self.event_running = True
+        self._exec_gcode(self.autoload_gcode)
+        self.event_running = False
+    def _exec_gcode(self, script):
+        try:
+            self.gcode.run_script(script)
+        except self.gcode.error as err:
+            # We arent inside a gcode handler, so respond with an error
+            self.gcode.respond_error(str(err))
+        except:
+            raise
+    def set_callbacks(self, runout_cb=None, detected_cb=None, monitor=True):
+        # Allow other modules to receive sensor status notifications
```

Comment:
I don't understand what this method does - it isn't invoked anywhere.  If the goal is to allow other modules to be aware of this event, then I'd use `self.printer.send_event("filament_switch_sensor:my_event", some, parameters)`

Reply:
Yes, this method is specifically for external modules to be able to set their own callbacks.  I have some concerns with using Klippers internal event system for this, as it would execute a callback for every module registered.   I'll provide an example:

An MMU module is changing filament, and it wants to be notified when filament triggers the sensor closest to the extruder.   To do this it would have to change the behavior of the sensor, as it would need to enable insert detection while the printer is in a "printing" state.  Also, it wouldn't be desirable for the built-in callback to execute the autoload_gcode while its being used for this.

There are ways around this.  Let me think on this and I'll see if I can come up with a solution.  I do think that I would need a more generic name for the event prefix, I wouldnt want to have to register event handlers for each underlying implementation (ie: filament_switch_sensor:runout_event, pat9125:runout_event, etc)

Reply:
FWIW, I'd say add that callback support when the software for the first callee is added.  It's hard to predict what future code will need.